Enable more compiler warnings

sledgehammer999 opened the pull request:
Also set warning level for MSVC
Affects CMake build system

NOTE: Making a draft because I suspect we will get enough warnings that will stall merging (as errors).

Chocobo1 commented on .github/workflows/ci_macos.yaml:
```diff
@@ -86,7 +86,7 @@ jobs:
         if: ${{ startsWith(matrix.qt_version, 5) }}
         run: |
-          CXXFLAGS="$CXXFLAGS -Werror -Wno-error=deprecated-declarations" \
+          CXXFLAGS="$CXXFLAGS -Wall -Wextra -Wpedantic -Wshadow -Werror -Wno-error=deprecated-declarations" \
```
It is already done in build scripts. FYI
https://github.com/qbittorrent/qBittorrent/blob/8df80b67f9f469a3240856f326d04da0ea221b11/cmake/Modules/CommonConfig.cmake#L51-L63
https://github.com/qbittorrent/qBittorrent/blob/8df80b67f9f469a3240856f326d04da0ea221b11/unixconf.pri#L13

sledgehammer999 replied:
Thanks. I have uploaded a new diff.

sledgehammer999 commented:
Update:

I added compiler warning level for MSVC.
I added warning about shadowing variables. I think shadowing is bad. It might allow the programmer to be confused on which variable he actually operates on in a given scope.
This affects the CMake build system.

Thoughts?

glassez commented:
@sledgehammer999 
How about 3rd-party code shipped with qBittorrent?

xavier2k6 commented:
@sledgehammer999
Would `/sdl` (Enable Additional Security Checks) be of any benefit also for MSVC?
https://learn.microsoft.com/en-us/cpp/build/reference/sdl-enable-additional-security-checks?view=msvc-170

sledgehammer999 commented:
>How about 3rd-party code shipped with qBittorrent?

Currently there is only one header file.
Regardless of that I have 2 proposals:

### 1st proposal
Wrapper header of 3rd party header:
```c++
#pragma once
// pseudo pragma below
#pragma disable warnings
#include <third/party/header.h>
// pseudo pragma below
#pragma enable warnings
```

qbittorrent .h/.cpp file:
```c++
#include <wraped_header.h>

// do stuff
```

### 2nd proposal
qbittorrent .h/.cpp file:
```c++
// regular includes go here

// 3rd party header includes
// pseudo pragma below
#pragma disable warnings
#include <third/party/header1.h>
#include <third/party/header2.h>
#include <third/party/header3.h>
// pseudo pragma below
#pragma enable warnings

// do stuff
```

NOTE: The pragmas that enable and disable warnings are different per compiler. They can be collected into one header for disabling warning and into a second header enabling them.

glassez commented:
>Regardless of that I have 2 proposals:

I don't like both proposals, since it mixes the build configuration related stuff into the source code (although the first option is to a lesser extent).
Is there no way to ignore such files from within build scripts?

glassez commented:
> Is there no way to ignore such files from within build scripts?

https://stackoverflow.com/questions/15133332/ignore-only-show-errors-warnings-from-certain-directory-using-cmake
Didn't read it carefully, but seems it is possible.

sledgehammer999 commented:
> > Is there no way to ignore such files from within build scripts?
> 
> https://stackoverflow.com/questions/15133332/ignore-only-show-errors-warnings-from-certain-directory-using-cmake Didn't read it carefully, but seems it is possible.

I don't think this will work. The 3rd party header is consumed by files that will have warnings enabled.

glassez commented:
> > How about 3rd-party code shipped with qBittorrent?
> 
> Currently there is only one header file. Regardless of that I have 2 proposals

Considering the above, we could use your first proposal, and wrap its header in `base/expected.h` (and I would also move `3rdparty` subfolder out from `base`.
